File: mobile/lib/data/entry_prefs.dart
import 'package:shared_preferences/shared_preferences.dart';

import 'models.dart';

/// Device-local UI preferences for the entry form: the last-used account and
/// payment method. These are conveniences, not financial data, so plain
/// SharedPreferences is fine (and this is not a local database/cache).
class EntryPrefs {
  EntryPrefs(this._prefs);

  final SharedPreferences _prefs;

  static const _accountKey = 'entry.lastAccountId';
  static const _methodKey = 'entry.lastPaymentMethod';

  String? get lastAccountId => _prefs.getString(_accountKey);

  PaymentMethod? get lastPaymentMethod => PaymentMethod.fromDb(_prefs.getString(_methodKey));

  Future<void> remember({required String accountId, PaymentMethod? method}) async {
    await _prefs.setString(_accountKey, accountId);
    if (method != null) await _prefs.setString(_methodKey, method.db);
  }
}

File: web/src/data/entryPrefs.ts
import type { PaymentMethod } from '@/lib/models'

/**
 * The last-used account and payment method, remembered in this browser so
 * the next entry starts with them (like the phone's EntryPrefs). These are
 * UI conveniences, not financial data, so localStorage is fine; this is not
 * a local database or cache.
 */
export interface EntryPrefs {
  readonly lastAccountId: string | null
  readonly lastMethod: PaymentMethod | null
  remember(accountId: string, method: PaymentMethod | null): void
}

const ACCOUNT_KEY = 'ventrafin.entry.lastAccountId'
const METHOD_KEY = 'ventrafin.entry.lastPaymentMethod'

function read(key: string): string | null {
  try {
    return localStorage.getItem(key)
  } catch {
    return null
  }
}

function write(key: string, value: string): void {
  try {
    localStorage.setItem(key, value)
  } catch {
    // Storage blocked (private window): the defaults just aren't remembered.
  }
}

export function createLocalEntryPrefs(): EntryPrefs {
  return {
    get lastAccountId() {
      return read(ACCOUNT_KEY)
    },
    get lastMethod() {
      const m = read(METHOD_KEY)
      return m === 'cash' || m === 'upi' || m === 'debit' || m === 'card' ? m : null
    },
    remember(accountId, method) {
      write(ACCOUNT_KEY, accountId)
      if (method) write(METHOD_KEY, method)
    },
  }
}
